Introduce `gh agent-task create`

BagToad opened the pull request:
Introduces the 'create' subcommand for agent tasks, allowing users to create agent tasks by providing a problem description as an argument, with the agent session web URL being return upon successful agent task creation.

BagToad committed:
Add agent task creation command and job API

Introduces the 'create' subcommand for agent tasks, allowing users to create agent jobs via the Copilot API. Adds job API client methods, job model, and polling logic to retrieve associated pull requests. Includes tests for various job creation scenarios.

babakks commented on pkg/cmd/agent-task/capi/job.go:
```diff
@@ -0,0 +1,122 @@
+		return nil, errors.New("problem statement is required")
+	}
+
+	url := fmt.Sprintf("%s/%s/%s", jobsBasePathV1, owner, repo)
```
Let's make these args path-safe:

```suggestion
	url := fmt.Sprintf("%s/%s/%s", url.PathEscape(jobsBasePathV1), url.PathEscape(owner), url.PathEscape(repo))
```

BagToad replied:
b890f880c0ed6a5f984735a6e6bc2cc43b542906

babakks commented on pkg/cmd/agent-task/capi/job.go:
```diff
@@ -0,0 +1,122 @@
+		"problem_statement": problemStatement,
+		"event_type":        defaultEventType,
+	}
+	b, _ := json.Marshal(body)
```
Why we're not using the `Job` type instead of a plain map/body?

BagToad replied:
Fixed in 5472fcffe9a41ae84f8a7d0c216ddd13e17c5cc9

babakks commented on pkg/cmd/agent-task/capi/job.go:
```diff
@@ -0,0 +1,122 @@
+		ID string `json:"id"`
+	} `json:"workflow_run,omitempty"`
+	ErrorInfo *JobError `json:"error,omitempty"`
+}
```
Seems like the `event_type` is missing on this type.

BagToad replied:
Fixed 32bf9159ea429de66c8130760009c5343e3ca711

babakks commented on pkg/cmd/agent-task/capi/job.go:
```diff
@@ -0,0 +1,122 @@
+				Message string `json:"message"`
+			} `json:"error"`
+		}
+		_ = json.NewDecoder(res.Body).Decode(&er)
```
Can't we use the `JobError` type here?

BagToad replied:
Really good idea. This simplifies this quite a bit since `JobError` is already part of the `Job` type and gets decoded when we decode the response.

Fixed in 21ccabc832f1c8e2ea88e7fb9678815e97dc330a

BagToad replied:
Fixed

babakks commented on pkg/cmd/agent-task/capi/job.go:
```diff
@@ -0,0 +1,122 @@
+	if owner == "" || repo == "" || jobID == "" {
+		return nil, errors.New("owner, repo, and jobID are required")
+	}
+	url := fmt.Sprintf("%s/%s/%s/%s", jobsBasePathV1, owner, repo, jobID)
```
Same comment:

```suggestion
	url := fmt.Sprintf("%s/%s/%s/%s", url.PathEscape(jobsBasePathV1), url.PathEscape(owner), url.PathEscape(repo), url.PathEscape(jobID))
```

BagToad replied:
9ec10a4d318c2d4f68bd60959862c34e457e7a4e

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+		IO: f.IOStreams,
+	}
+	cmd := &cobra.Command{
+		Use:   "create <problem statement>",
```
The `<problem statement>` looks a bit weird to me. How about `<task description>`?

BagToad replied:
I agree. So internally we'll call it problem statement because that's what the API calls it, and then publicly we'll call it task description.

I also added quotes to the usage - not sure if I should though? Let me know what you think?

Fixed in a81cff3fdfddb21adb8c544d14f70a442440e565

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+	}
+	cmd := &cobra.Command{
+		Use:   "create <problem statement>",
+		Short: "Create an agent task (preview)",
```
We should either add `(preview)` to all commands, or just add it once (to the base cmd).

BagToad replied:
I feel like it probably should be in the help of every command including the base command.

Ultimately whatever it takes to show up in each sub-command's help text because nobody will probably look at the base command's help more than once.

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+		Short: "Create an agent task (preview)",
+		Args:  cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.ProblemStatement = strings.Join(args, " ")
```
This can cause confusion and unexpected results, I guess.

As an example, running this will return a repo error, while the user meant something else.

```sh
gh agent-task create add --repo option to command
```

I know the user can always use the stop-parsing arg (i.e. `--`), but still a single (quoted) arg makes more sense.

What do you think?

BagToad replied:
Thanks - great point.

Fixed in a3fa83071d791e024c76b8f52fec902a51272fa1

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+		return errors.New("failed to resolve repository")
+	}
+	repo, err := opts.BaseRepo()
+	if err != nil || repo == nil || repo.RepoOwner() == "" || repo.RepoName() == "" {
```
I think it's a bit extra:


```suggestion
	if err != nil || repo == nil {
```

BagToad replied:
Fixed this and a few other things in a821b408d41db3113a56bfd10666a649e604e494

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+	if err != nil || repo == nil || repo.RepoOwner() == "" || repo.RepoName() == "" {
+		// Not printing the error that came back from BaseRepo() here because we want
+		// something clear, human friendly, and actionable.
+		return fmt.Errorf("error: a repository is required; re-run in a repository or supply one with --repo owner/name")
```
Maybe it's better to remove the `error:` prefix:

```suggestion
		return fmt.Errorf("a repository is required; re-run in a repository or supply one with --repo owner/name")
```

BagToad replied:
3a7465ed96c5206f423cf33bfc7e7b95f823ffc1

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+func agentSessionWebURL(repo ghrepo.Interface, j *capi.Job) string {
+	if j == nil || j.PullRequest == nil {
```
This func is called from two places both of which provide a non-`nil` job argument. So, we can drop the `nil` check and let the potential bugs surface with a panic.


```suggestion
	if j.PullRequest == nil {
```

BagToad replied:
c3bbd374aa97444298f7cd76d017919b0bd4fcea

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+	if j == nil || j.PullRequest == nil {
+		return ""
+	}
+	if j.SessionID == "" {
+		return fmt.Sprintf("https://github.com/%s/%s/pull/%d", repo.RepoOwner(), repo.RepoName(), j.PullRequest.Number)
+	}
+	return fmt.Sprintf("https://github.com/%s/%s/pull/%d/agent-sessions/%s", repo.RepoOwner(), repo.RepoName(), j.PullRequest.Number, j.SessionID)
+}
```
We should apply `url.PathEscape` to path components.

BagToad replied:
33d119664525dc2f7fc78dab68c45507b24355f1

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+	var errPRNotReady = errors.New("job not ready")
+
+	var result *capi.Job
+	retryErr := backoff.Retry(func() error {
+		j, getErr := client.GetJob(ctx, repo.RepoOwner(), repo.RepoName(), jobID)
+		if getErr != nil {
+			fmt.Fprintf(errOut, "warning: failed to get job status: %v\n", getErr)
+			return errPRNotReady
+		}
+		if j.PullRequest != nil && j.PullRequest.Number > 0 {
+			result = j
+			return nil
+		}
+		return errPRNotReady
+	}, backoff.WithContext(bo, ctx))
```
Shouldn't we stop retrying in the first `if`? Because we know the job is created and we expect the `GET` request to work? If we do that, we don't even need to pass the stream writer.

What do you think?

In case you agree with the change, here's a suggestion:


```suggestion
	retryErr := backoff.Retry(func() error {
		j, err := client.GetJob(ctx, repo.RepoOwner(), repo.RepoName(), jobID)
		if err != nil {
			return backoff.Permanent(err)
		}
		if j.PullRequest != nil && j.PullRequest.Number > 0 {
			result = j
			return nil
		}
		return errPRNotReady
	}, backoff.WithContext(bo, ctx))
```

BagToad replied:
32bf9159ea429de66c8130760009c5343e3ca711

babakks commented on pkg/cmd/agent-task/create/create.go:
```diff
@@ -0,0 +1,173 @@
+
+	if retryErr != nil {
+		if errors.Is(retryErr, errPRNotReady) {
+			// Timed out or failed to fetch
```
If we changed the error handling, then this comment should be just `// Timed out`.

BagToad replied:
32bf9159ea429de66c8130760009c5343e3ca711

babakks commented on pkg/cmd/agent-task/create/create_test.go:
```diff
@@ -0,0 +1,191 @@
+			stubs: func(reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.WithHost(httpmock.REST("POST", "agents/swe/v1/jobs/OWNER/REPO"), "api.githubcopilot.com"),
+					httpmock.StatusStringResponse(400, `{"error":{"message":"some API error"}}`),
```
Maybe `HTTP 500` instead of `400`?

BagToad replied:
3d8d5f3e312321b4993583b694a0468ffcd2da60

babakks reviewed (approved):
Thanks for the PR! Looks very good. 🙏

BagToad committed:
Update agent-task create command argument handling

Changed argument parsing to accept a maximum of one argument and assign it directly to ProblemStatement. Added error handling for missing problem statement and removed unused strings import.

BagToad committed:
Update command usage for agent task creation

Changed the command usage string from 'create <problem statement>' to 'create "<task description>"' for improved clarity in the agent task creation command.

BagToad committed:
Remove redundant 'error:' prefix from repo error

Simplifies the error message when a repository is missing by removing the unnecessary 'error:' prefix, making the output clearer and more user-friendly.

BagToad committed:
Update error message in createRun test

Removed 'error:' prefix from expected error message in the 'missing repo returns error' test case for consistency with actual output.

BagToad committed:
Update test to expect 500 error response

Changed the mocked API response status from 400 to 500 in Test_createRun to better reflect the expected error scenario.

BagToad committed:
Remove nil check for job in agentSessionWebURL

Simplifies the agentSessionWebURL function by removing the redundant nil check for the job parameter, assuming it is always non-nil when called.

BagToad committed:
Escape URL path segments in agent session links

Uses url.PathEscape for repo owner, repo name, and session ID when constructing agent session URLs to ensure proper encoding and prevent issues with special characters.

BagToad committed:
Escape owner and repo in job creation URL

Uses url.PathEscape for owner and repo when constructing the job creation URL to ensure proper encoding and prevent issues with special characters.

BagToad committed:
Escape URL path parameters in GetJob request

Uses url.PathEscape for owner, repo, and jobID in the GetJob API request to ensure proper encoding and prevent issues with special characters.

BagToad committed:
Use Job struct in request

Refactored CreateJob to use the Job struct as the payload instead of a map, improving consistency and maintainability.

BagToad committed:
Improve job error handling and update tests

Normalizes job API error messages to include status code and text, ensures errors from job polling are surfaced to stderr without halting execution, and updates tests to verify error output and remove outdated fallback behavior.

BagToad committed:
Simplify stdout assertion in createRun test

Removed unnecessary conditional check for wantStdout in Test_createRun. Now always asserts equality between wantStdout and actual stdout output, improving test clarity.

BagToad committed:
Simplify error handling in CreateJob response

Refactored the CreateJob method to decode the response body into the Job struct before checking for error status codes. Error messages are now extracted directly from the Job struct, removing redundant error parsing logic.

BagToad committed:
Update error messages and test repo handling in agent-task create

Replaces 'problem statement' with 'task description' in error messages for clarity. Refactors tests to use a BaseRepo function instead of direct repo objects, and adds a test for missing task description error.
